os/drivers/seclink/seclink_drv_common.c

```diff
@@ -0,0 +1,60 @@
+/****************************************************************************
+ *
+ * Copyright 2019 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+#include <tinyara/config.h>
+
+#include <stdio.h>
+#include <tinyara/seclink.h>
+#include <tinyara/seclink_drv.h>
+#include "seclink_drv_req.h"
+#include "seclink_drv_utils.h"
+
+/*  Debugging */
+#ifdef SLDRV_TAG
+#undef SLDRV_TAG
+#endif
+#define SLDRV_TAG "[SECLINK_DRV_COM]"
+
+
+int hd_handle_common_request(int cmd, unsigned long arg, void *lower)
+{
+	SLDRV_ENTER;
+
+	struct seclink_req *req = (struct seclink_req *)arg;
+	hal_init_param *params = NULL;
+
+	if (!req) {
+		return -1;
+	}
+
+	struct sec_lowerhalf_s *se = (struct sec_lowerhalf_s *)lower;
+	if (!se | !(se->ops)) {
+		return -1;
+	}
+
+	switch (cmd) {
+	case SECLINK_HAL_INIT:
```

Comment:
It initialized i2c or gpio port.
This is very tightly with h/w, you had better define dev structure of each lower half device which using i2c or gpio.
```
FAR struct se_lowerhalf_s *A_initialize(FAR struct i2c_dev_s *i2c, FAR struct A_lower_s *lower) {
     priv = (FAR struct A_dev_s *)malloc();
     priv->lower = lower;
     priv->i2c = i2c;
FAR struct se_lowerhalf_s *B_initialize(FAR struct gpio_dev_s *gpio, FAR struct B_lower_s *lower) {
     priv = (FAR struct B_dev_s *)malloc();
     priv->lower = lower;
     priv->gpio = gpio;
}
```
Make sure that upper doesn't have to know lower device's lower device.

Reply:
basic idea is let se know what gpio or i2c can they use. so open or close the peripheral is their responsibility.

Reply:
It is same story of my other comment, gpio or i2c are tightly related to board, upper doesn't have to know lower's lower.
```
// seclink_drv_common
function a() {
    lower->ops->a(lower);
}
// lower of seclink_drv_common
function a() {
   FAR struct i2c_dev_s *dev;
   dev = priv->i2c;
   i2c_write(dev);
}

function a'() {
   FAR struct gpio_dev_s *dev;
   dev = priv->gpio;
   gpio_write(dev);
}

FAR struct lower_half a_initialize(FAR struct i2c_dev_s i2c or gpio_dev_s gpio) {
    FAR struct a_dev_s *priv;
    priv = malloc();
    priv->i2c = i2c;
    or
    priv->gpio = gpio;
}